# Day 37 — Logistic Regression

Linear Regression → Predicts continuous values (House Price, Salary)

Logistic Regression → Predicts categories (Pass/Fail, Yes/No, 0/1)

Despite the name "Regression", Logistic Regression is mainly used for Classification.

## Problem Statement
#### Predict whether a student will Pass (1) or Fail (0) based on study hours.

In [43]:
# Step 1: Create Dataset

import pandas as pd

data = {
    "Hours": [1,2,3,4,5,6,7,8,9,10],
    "Pass":  [0,0,0,0,0,1,1,1,1,1]
}

df = pd.DataFrame(data)

In [45]:
# Step 2: Features and Target

X = df[['Hours']]
y = df['Pass']

In [47]:
# Step 3: Train-Test Split

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [48]:
# Step 4: Train Logistic Regression

from sklearn.linear_model import LogisticRegression

model = LogisticRegression()

model.fit(X_train, y_train)

LogisticRegression()

In [49]:
 # Step 5: Prediction

y_pred = model.predict(X_test)

print(y_pred)

[1 0]


In [54]:
# Step 6: Accuracy

from sklearn.metrics import accuracy_score

acc = accuracy_score(y_test, y_pred)

print("Accuracy:", acc)

Accuracy: 1.0


In [55]:
# Step 7: Probability Prediction

print(model.predict_proba([[5]]))

[[0.63908874 0.36091126]]


C:\DevTools\Anaconda\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [59]:
# tep 8: Predict New Student

new_student = pd.DataFrame({
    'Hours': [7]
})

print(model.predict(new_student))

[1]


In [58]:
# Self Practice

X = df[['Hours']]
y = df['Pass']

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

from sklearn.linear_model import LogisticRegression 
model = LogisticRegression()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)
print(y_pred)

from sklearn.metrics import accuracy_score

acc = accuracy_score(y_pred, y_test)
print(f"Accuracy Score : {acc}")

probability = model.predict_proba([[9]])
print(f"Probability : {probability}")

# 2 % Fail
# 98 % Pass

[1 0]
Accuracy Score : 1.0
Probability : [[0.01799637 0.98200363]]


C:\DevTools\Anaconda\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


### New Concept: Sigmoid Function

## New Concept: Sigmoid Function

### Logistic Regression converts output into a probability between : 0 and 1 using the Sigmoid Function.

**Example:**
* 0.10 → Fail
* 0.90 → Pass

## Interview Questions
### After running the project, answer:

## 1. Why is Logistic Regression used for classification?
#### Logistic Regression is used for classification because it predicts the probability of belonging to a class and then assigns a category (0 or 1).

## 2. What is the output range of the Sigmoid Function?
#### Between 0 and 1.

## 3. What does predict_proba() return?
#### It predicts the probability for pass and fail.

* Example : [[0.02, 0.98]]
* means:

* 2% probability of Fail
* 98% probability of Pass

## 4. If probability = 0.9, what class will likely be predicted?
#### 90 percent chances of pass.

## 5. Difference between Linear Regression and Logistic Regression?

| Linear Regression          | Logistic Regression                |
| -------------------------- | ---------------------------------- |
| Predicts continuous values | Predicts categories                |
| House Price                | Pass/Fail                          |
| Salary                     | Spam/Not Spam                      |
| Output can be any number   | Output probability between 0 and 1 |
